File: .config/ai-sync/sync/sync.test.ts
import { describe, test, expect, beforeEach, afterEach } from "bun:test";
import { mkdirSync, rmSync, writeFileSync, existsSync, lstatSync, readlinkSync, symlinkSync } from "fs";
import { join } from "path";
import { tmpdir } from "os";
import { sync, AGENT_SKILL_PATHS, type Config } from "./sync";

describe("sync", () => {
  let tempDir: string;
  let skillsSourceDir: string;
  let projectDir: string;
  let configPath: string;

  beforeEach(() => {
    // Create unique temp directory for each test
    tempDir = join(tmpdir(), `ai-sync-test-${Date.now()}-${Math.random().toString(36).slice(2)}`);
    skillsSourceDir = join(tempDir, "skills");
    projectDir = join(tempDir, "test-project");
    configPath = join(tempDir, "config.json");

    // Create directory structure
    mkdirSync(skillsSourceDir, { recursive: true });
    mkdirSync(projectDir, { recursive: true });
  });

  afterEach(() => {
    // Cleanup temp directory
    if (existsSync(tempDir)) {
      rmSync(tempDir, { recursive: true, force: true });
    }
  });

  test("syncs skills to project directory for all agents", async () => {
    // Create mock skills
    const skillName = "test-skill";
    const skillDir = join(skillsSourceDir, skillName);
    mkdirSync(skillDir);
    writeFileSync(join(skillDir, "SKILL.md"), "# Test Skill");

    // Create config
    const config: Config = {
      projects: [
        {
          path: projectDir,
          skills: [skillName],
          agents: ["claude", "opencode"],
        },
      ],
    };
    writeFileSync(configPath, JSON.stringify(config, null, 2));

    // Run sync
    await sync({ configPath, skillsSourceDir });

    // Verify symlinks were created for both agents
    for (const agent of ["claude", "opencode"] as const) {
      const expectedSymlinkPath = join(projectDir, AGENT_SKILL_PATHS[agent], skillName);
      
      expect(existsSync(expectedSymlinkPath)).toBe(true);
      
      const stat = lstatSync(expectedSymlinkPath);
      expect(stat.isSymbolicLink()).toBe(true);
      
      const linkTarget = readlinkSync(expectedSymlinkPath);
      expect(linkTarget).toBe(skillDir);
    }
  });

  test("creates skill directories if they don't exist", async () => {
    const skillName = "another-skill";
    const skillDir = join(skillsSourceDir, skillName);
    mkdirSync(skillDir);
    writeFileSync(join(skillDir, "SKILL.md"), "# Another Skill");

    const config: Config = {
      projects: [
        {
          path: projectDir,
          skills: [skillName],
          agents: ["claude"],
        },
      ],
    };
    writeFileSync(configPath, JSON.stringify(config, null, 2));

    // Verify skill dir doesn't exist yet
    const skillTargetDir = join(projectDir, AGENT_SKILL_PATHS.claude);
    expect(existsSync(skillTargetDir)).toBe(false);

    await sync({ configPath, skillsSourceDir });

    // Verify it was created
    expect(existsSync(skillTargetDir)).toBe(true);
  });

  test("skips syncing if skill source doesn't exist", async () => {
    const config: Config = {
      projects: [
        {
          path: projectDir,
          skills: ["nonexistent-skill"],
          agents: ["claude"],
        },
      ],
    };
    writeFileSync(configPath, JSON.stringify(config, null, 2));

    // Should not throw
    await sync({ configPath, skillsSourceDir });

    // Skill directory should exist but be empty (no symlinks)
    const skillTargetDir = join(projectDir, AGENT_SKILL_PATHS.claude);
    expect(existsSync(skillTargetDir)).toBe(true);
    expect(existsSync(join(skillTargetDir, "nonexistent-skill"))).toBe(false);
  });

  test("handles projects with no skills configured", async () => {
    const config: Config = {
      projects: [
        {
          path: projectDir,
          skills: [],
          agents: ["claude"],
        },
      ],
    };
    writeFileSync(configPath, JSON.stringify(config, null, 2));

    // Should not throw
    await sync({ configPath, skillsSourceDir });
  });

  test("skips already linked skills", async () => {
    const skillName = "existing-skill";
    const skillDir = join(skillsSourceDir, skillName);
    mkdirSync(skillDir);
    writeFileSync(join(skillDir, "SKILL.md"), "# Existing");

    const config: Config = {
      projects: [
        {
          path: projectDir,
          skills: [skillName],
          agents: ["claude"],
        },
      ],
    };
    writeFileSync(configPath, JSON.stringify(config, null, 2));

    // Run sync twice
    await sync({ configPath, skillsSourceDir });
    await sync({ configPath, skillsSourceDir });

    // Should still work, symlink should still exist
    const symlinkPath = join(projectDir, AGENT_SKILL_PATHS.claude, skillName);
    expect(existsSync(symlinkPath)).toBe(true);
    expect(lstatSync(symlinkPath).isSymbolicLink()).toBe(true);
  });

  test("handles multiple skills", async () => {
    const skills = ["skill-one", "skill-two", "skill-three"];
    
    for (const skill of skills) {
      const skillDir = join(skillsSourceDir, skill);
      mkdirSync(skillDir);
      writeFileSync(join(skillDir, "SKILL.md"), `# ${skill}`);
    }

    const config: Config = {
      projects: [
        {
          path: projectDir,
          skills,
          agents: ["opencode"],
        },
      ],
    };
    writeFileSync(configPath, JSON.stringify(config, null, 2));

    await sync({ configPath, skillsSourceDir });

    for (const skill of skills) {
      const symlinkPath = join(projectDir, AGENT_SKILL_PATHS.opencode, skill);
      expect(existsSync(symlinkPath)).toBe(true);
      expect(lstatSync(symlinkPath).isSymbolicLink()).toBe(true);
    }
  });

  test("handles multiple projects", async () => {
    const secondProjectDir = join(tempDir, "second-project");
    mkdirSync(secondProjectDir);

    const skillName = "shared-skill";
    const skillDir = join(skillsSourceDir, skillName);
    mkdirSync(skillDir);
    writeFileSync(join(skillDir, "SKILL.md"), "# Shared");

    const config: Config = {
      projects: [
        {
          path: projectDir,
          skills: [skillName],
          agents: ["claude"],
        },
        {
          path: secondProjectDir,
          skills: [skillName],
          agents: ["opencode"],
        },
      ],
    };
    writeFileSync(configPath, JSON.stringify(config, null, 2));

    await sync({ configPath, skillsSourceDir });

    // Verify first project
    expect(existsSync(join(projectDir, AGENT_SKILL_PATHS.claude, skillName))).toBe(true);
    
    // Verify second project
    expect(existsSync(join(secondProjectDir, AGENT_SKILL_PATHS.opencode, skillName))).toBe(true);
  });

  test("removes orphaned symlinks for skills no longer in config", async () => {
    // Create two skills
    const skillToKeep = "keep-skill";
    const skillToRemove = "remove-skill";
    
    for (const skill of [skillToKeep, skillToRemove]) {
      const skillDir = join(skillsSourceDir, skill);
      mkdirSync(skillDir);
      writeFileSync(join(skillDir, "SKILL.md"), `# ${skill}`);
    }

    // Initial config with both skills
    const config: Config = {
      projects: [
        {
          path: projectDir,
          skills: [skillToKeep, skillToRemove],
          agents: ["claude"],
        },
      ],
    };
    writeFileSync(configPath, JSON.stringify(config, null, 2));

    // Run sync to create both symlinks
    await sync({ configPath, skillsSourceDir });

    // Verify both symlinks exist
    const keepPath = join(projectDir, AGENT_SKILL_PATHS.claude, skillToKeep);
    const removePath = join(projectDir, AGENT_SKILL_PATHS.claude, skillToRemove);
    expect(existsSync(keepPath)).toBe(true);
    expect(existsSync(removePath)).toBe(true);

    // Update config to only include one skill
    config.projects[0].skills = [skillToKeep];
    writeFileSync(configPath, JSON.stringify(config, null, 2));

    // Run sync again
    await sync({ configPath, skillsSourceDir });

    // Verify keepPath still exists
    expect(existsSync(keepPath)).toBe(true);
    expect(lstatSync(keepPath).isSymbolicLink()).toBe(true);

    // Verify removePath was cleaned up
    expect(existsSync(removePath)).toBe(false);
  });

  test("does not remove symlinks pointing to other directories", async () => {
    const skillName = "my-skill";
    const skillDir = join(skillsSourceDir, skillName);
    mkdirSync(skillDir);
    writeFileSync(join(skillDir, "SKILL.md"), "# My Skill");

    // Create a symlink pointing to a different directory (not our skills source)
    const skillsTargetDir = join(projectDir, AGENT_SKILL_PATHS.claude);
    mkdirSync(skillsTargetDir, { recursive: true });
    
    const externalDir = join(tempDir, "external");
    mkdirSync(externalDir);
    const externalSymlink = join(skillsTargetDir, "external-skill");
    symlinkSync(externalDir, externalSymlink);

    const config: Config = {
      projects: [
        {
          path: projectDir,
          skills: [skillName],
          agents: ["claude"],
        },
      ],
    };
    writeFileSync(configPath, JSON.stringify(config, null, 2));

    await sync({ configPath, skillsSourceDir });

    // The external symlink should still exist (not cleaned up)
    expect(existsSync(externalSymlink)).toBe(true);
    expect(lstatSync(externalSymlink).isSymbolicLink()).toBe(true);

    // Our skill should be linked
    expect(existsSync(join(skillsTargetDir, skillName))).toBe(true);
  });

  test("does not remove regular files or directories during cleanup", async () => {
    const skillName = "my-skill";
    const skillDir = join(skillsSourceDir, skillName);
    mkdirSync(skillDir);
    writeFileSync(join(skillDir, "SKILL.md"), "# My Skill");

    // Create regular file and directory in the skills target
    const skillsTargetDir = join(projectDir, AGENT_SKILL_PATHS.claude);
    mkdirSync(skillsTargetDir, { recursive: true });
    
    const regularFile = join(skillsTargetDir, "regular-file.txt");
    writeFileSync(regularFile, "content");
    
    const regularDir = join(skillsTargetDir, "regular-dir");
    mkdirSync(regularDir);

    const config: Config = {
      projects: [
        {
          path: projectDir,
          skills: [skillName],
          agents: ["claude"],
        },
      ],
    };
    writeFileSync(configPath, JSON.stringify(config, null, 2));

    await sync({ configPath, skillsSourceDir });

    // Regular file and directory should still exist
    expect(existsSync(regularFile)).toBe(true);
    expect(existsSync(regularDir)).toBe(true);
    expect(lstatSync(regularFile).isFile()).toBe(true);
    expect(lstatSync(regularDir).isDirectory()).toBe(true);
  });
});

File: .config/ai-sync/cmd/sync/sync_test.go
package main

import (
	"os"
	"path/filepath"
	"testing"

	"github.com/stretchr/testify/require"
	"gopkg.in/yaml.v3"
)

type testEnv struct {
	tempDir           string
	skillsSourceDir   string
	commandsSourceDir string
	agentsSourceDir   string
	projectDir        string
	configPath        string
}

func setupTestEnv(t *testing.T) *testEnv {
	t.Helper()

	tempDir := t.TempDir()

	env := &testEnv{
		tempDir:           tempDir,
		skillsSourceDir:   filepath.Join(tempDir, "skills"),
		commandsSourceDir: filepath.Join(tempDir, "commands"),
		agentsSourceDir:   filepath.Join(tempDir, "agents"),
		projectDir:        filepath.Join(tempDir, "test-project"),
		configPath:        filepath.Join(tempDir, "config.yaml"),
	}

	require.NoError(t, os.MkdirAll(env.skillsSourceDir, 0755))
	require.NoError(t, os.MkdirAll(env.commandsSourceDir, 0755))
	require.NoError(t, os.MkdirAll(env.agentsSourceDir, 0755))
	require.NoError(t, os.MkdirAll(env.projectDir, 0755))

	return env
}

func (e *testEnv) writeConfig(t *testing.T, config *Config) {
	t.Helper()
	data, err := yaml.Marshal(config)
	require.NoError(t, err)
	require.NoError(t, os.WriteFile(e.configPath, data, 0644))
}

func (e *testEnv) createSkill(t *testing.T, name string) {
	t.Helper()
	skillDir := filepath.Join(e.skillsSourceDir, name)
	require.NoError(t, os.MkdirAll(skillDir, 0755))
	require.NoError(t, os.WriteFile(filepath.Join(skillDir, "SKILL.md"), []byte("# "+name), 0644))
}

func (e *testEnv) createCommand(t *testing.T, name string) {
	t.Helper()
	require.NoError(t, os.WriteFile(filepath.Join(e.commandsSourceDir, name+".md"), []byte("# "+name), 0644))
}

func (e *testEnv) createAgent(t *testing.T, name string) {
	t.Helper()
	require.NoError(t, os.WriteFile(filepath.Join(e.agentsSourceDir, name+".md"), []byte("# "+name), 0644))
}

func (e *testEnv) syncOpts() SyncOptions {
	return SyncOptions{
		ConfigPath:        e.configPath,
		SkillsSourceDir:   e.skillsSourceDir,
		CommandsSourceDir: e.commandsSourceDir,
		AgentsSourceDir:   e.agentsSourceDir,
	}
}

func TestSyncSkillsCommandsAgents(t *testing.T) {
	env := setupTestEnv(t)

	// Create test items
	env.createSkill(t, "test-skill")
	env.createCommand(t, "test-command")
	env.createAgent(t, "test-agent")

	env.writeConfig(t, &Config{
		Projects: []ProjectConfig{
			{
				Path: env.projectDir,
				AiStuff: &AiStuff{
					Skills:       []string{"test-skill"},
					Commands:     []string{"test-command"},
					CustomAgents: []string{"test-agent"},
				},
				Harnesses: []Harness{HarnessClaude, HarnessOpencode},
			},
		},
	})

	require.NoError(t, Sync(env.syncOpts()))

	// Verify symlinks for both harnesses
	for _, harness := range []Harness{HarnessClaude, HarnessOpencode} {
		// Check skill symlink
		skillPath := filepath.Join(env.projectDir, HarnessSkillPaths[harness], "test-skill")
		info, err := os.Lstat(skillPath)
		require.NoError(t, err, "skill symlink should exist for %s", harness)
		require.True(t, info.Mode()&os.ModeSymlink != 0, "should be a symlink")

		target, err := os.Readlink(skillPath)
		require.NoError(t, err)
		require.Equal(t, filepath.Join(env.skillsSourceDir, "test-skill"), target)

		// Check command symlink
		cmdPath := filepath.Join(env.projectDir, HarnessCommandPaths[harness], "test-command.md")
		info, err = os.Lstat(cmdPath)
		require.NoError(t, err, "command symlink should exist for %s", harness)
		require.True(t, info.Mode()&os.ModeSymlink != 0, "should be a symlink")

		// Check agent symlink
		agentPath := filepath.Join(env.projectDir, HarnessAgentPaths[harness], "test-agent.md")
		info, err = os.Lstat(agentPath)
		require.NoError(t, err, "agent symlink should exist for %s", harness)
		require.True(t, info.Mode()&os.ModeSymlink != 0, "should be a symlink")
	}
}

func TestSyncCreatesDirectories(t *testing.T) {
	env := setupTestEnv(t)
	env.createSkill(t, "my-skill")

	// Verify target directories don't exist yet
	skillTargetDir := filepath.Join(env.projectDir, HarnessSkillPaths[HarnessClaude])
	_, err := os.Stat(skillTargetDir)
	require.True(t, os.IsNotExist(err), "directory should not exist before sync")

	env.writeConfig(t, &Config{
		Projects: []ProjectConfig{
			{
				Path:      env.projectDir,
				AiStuff:   &AiStuff{Skills: []string{"my-skill"}},
				Harnesses: []Harness{HarnessClaude},
			},
		},
	})

	require.NoError(t, Sync(env.syncOpts()))

	// Verify directory was created
	info, err := os.Stat(skillTargetDir)
	require.NoError(t, err)
	require.True(t, info.IsDir())
}

func TestSyncErrorsOnMissingSource(t *testing.T) {
	env := setupTestEnv(t)

	tests := []struct {
		name    string
		config  *Config
		errPart string
	}{
		{
			name: "missing skill",
			config: &Config{
				Projects: []ProjectConfig{
					{
						Path:      env.projectDir,
						AiStuff:   &AiStuff{Skills: []string{"nonexistent-skill"}},
						Harnesses: []Harness{HarnessClaude},
					},
				},
			},
			errPart: "Skill source not found",
		},
		{
			name: "missing command",
			config: &Config{
				Projects: []ProjectConfig{
					{
						Path:      env.projectDir,
						AiStuff:   &AiStuff{Commands: []string{"nonexistent-cmd"}},
						Harnesses: []Harness{HarnessClaude},
					},
				},
			},
			errPart: "Command source not found",
		},
		{
			name: "missing agent",
			config: &Config{
				Projects: []ProjectConfig{
					{
						Path:      env.projectDir,
						AiStuff:   &AiStuff{CustomAgents: []string{"nonexistent-agent"}},
						Harnesses: []Harness{HarnessClaude},
					},
				},
			},
			errPart: "Agent source not found",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			env.writeConfig(t, tt.config)
			err := Sync(env.syncOpts())
			require.Error(t, err)
			require.Contains(t, err.Error(), tt.errPart)
		})
	}
}

func TestSyncIdempotent(t *testing.T) {
	env := setupTestEnv(t)
	env.createSkill(t, "idem-skill")
	env.createCommand(t, "idem-cmd")

	env.writeConfig(t, &Config{
		Projects: []ProjectConfig{
			{
				Path: env.projectDir,
				AiStuff: &AiStuff{
					Skills:   []string{"idem-skill"},
					Commands: []string{"idem-cmd"},
				},
				Harnesses: []Harness{HarnessOpencode},
			},
		},
	})

	// Run sync twice
	require.NoError(t, Sync(env.syncOpts()))
	require.NoError(t, Sync(env.syncOpts()))

	// Verify symlinks still work
	skillPath := filepath.Join(env.projectDir, HarnessSkillPaths[HarnessOpencode], "idem-skill")
	info, err := os.Lstat(skillPath)
	require.NoError(t, err)
	require.True(t, info.Mode()&os.ModeSymlink != 0)

	cmdPath := filepath.Join(env.projectDir, HarnessCommandPaths[HarnessOpencode], "idem-cmd.md")
	info, err = os.Lstat(cmdPath)
	require.NoError(t, err)
	require.True(t, info.Mode()&os.ModeSymlink != 0)
}

func TestSyncCleansUpOrphans(t *testing.T) {
	env := setupTestEnv(t)

	// Create skills
	env.createSkill(t, "keep-skill")
	env.createSkill(t, "remove-skill")

	// Initial sync with both skills
	env.writeConfig(t, &Config{
		Projects: []ProjectConfig{
			{
				Path:      env.projectDir,
				AiStuff:   &AiStuff{Skills: []string{"keep-skill", "remove-skill"}},
				Harnesses: []Harness{HarnessClaude},
			},
		},
	})
	require.NoError(t, Sync(env.syncOpts()))

	keepPath := filepath.Join(env.projectDir, HarnessSkillPaths[HarnessClaude], "keep-skill")
	removePath := filepath.Join(env.projectDir, HarnessSkillPaths[HarnessClaude], "remove-skill")

	// Both should exist
	require.FileExists(t, keepPath)
	require.FileExists(t, removePath)

	// Update config to only include one skill
	env.writeConfig(t, &Config{
		Projects: []ProjectConfig{
			{
				Path:      env.projectDir,
				AiStuff:   &AiStuff{Skills: []string{"keep-skill"}},
				Harnesses: []Harness{HarnessClaude},
			},
		},
	})
	require.NoError(t, Sync(env.syncOpts()))

	// keep-skill should still exist
	require.FileExists(t, keepPath)

	// remove-skill should be cleaned up
	_, err := os.Lstat(removePath)
	require.True(t, os.IsNotExist(err), "orphaned symlink should be removed")

	// Test: external symlinks are NOT removed
	skillsTargetDir := filepath.Join(env.projectDir, HarnessSkillPaths[HarnessClaude])
	externalDir := filepath.Join(env.tempDir, "external")
	require.NoError(t, os.MkdirAll(externalDir, 0755))

	externalSymlink := filepath.Join(skillsTargetDir, "external-skill")
	require.NoError(t, os.Symlink(externalDir, externalSymlink))

	// Run sync again
	require.NoError(t, Sync(env.syncOpts()))

	// External symlink should still exist
	info, err := os.Lstat(externalSymlink)
	require.NoError(t, err, "external symlink should not be removed")
	require.True(t, info.Mode()&os.ModeSymlink != 0)

	// Test: regular files are NOT removed
	regularFile := filepath.Join(skillsTargetDir, "regular-file.txt")
	require.NoError(t, os.WriteFile(regularFile, []byte("content"), 0644))

	require.NoError(t, Sync(env.syncOpts()))

	info, err = os.Stat(regularFile)
	require.NoError(t, err, "regular file should not be removed")
	require.True(t, info.Mode().IsRegular())
}

func TestSyncMultipleProjects(t *testing.T) {
	env := setupTestEnv(t)

	secondProjectDir := filepath.Join(env.tempDir, "second-project")
	require.NoError(t, os.MkdirAll(secondProjectDir, 0755))

	env.createSkill(t, "shared-skill")

	env.writeConfig(t, &Config{
		Projects: []ProjectConfig{
			{
				Path:      env.projectDir,
				AiStuff:   &AiStuff{Skills: []string{"shared-skill"}},
				Harnesses: []Harness{HarnessClaude},
			},
			{
				Path:      secondProjectDir,
				AiStuff:   &AiStuff{Skills: []string{"shared-skill"}},
				Harnesses: []Harness{HarnessOpencode},
			},
		},
	})

	require.NoError(t, Sync(env.syncOpts()))

	// Verify first project (claude harness)
	path1 := filepath.Join(env.projectDir, HarnessSkillPaths[HarnessClaude], "shared-skill")
	info, err := os.Lstat(path1)
	require.NoError(t, err)
	require.True(t, info.Mode()&os.ModeSymlink != 0)

	// Verify second project (opencode harness)
	path2 := filepath.Join(secondProjectDir, HarnessSkillPaths[HarnessOpencode], "shared-skill")
	info, err = os.Lstat(path2)
	require.NoError(t, err)
	require.True(t, info.Mode()&os.ModeSymlink != 0)
}

func TestSyncEmptyConfig(t *testing.T) {
	env := setupTestEnv(t)

	// No projects
	env.writeConfig(t, &Config{Projects: []ProjectConfig{}})
	require.NoError(t, Sync(env.syncOpts()))

	// Project with no ai-stuff
	env.writeConfig(t, &Config{
		Projects: []ProjectConfig{
			{
				Path:      env.projectDir,
				AiStuff:   nil,
				Harnesses: []Harness{HarnessClaude},
			},
		},
	})
	require.NoError(t, Sync(env.syncOpts()))

	// Project with empty lists
	env.writeConfig(t, &Config{
		Projects: []ProjectConfig{
			{
				Path:      env.projectDir,
				AiStuff:   &AiStuff{Skills: []string{}, Commands: []string{}, CustomAgents: []string{}},
				Harnesses: []Harness{HarnessClaude},
			},
		},
	})
	require.NoError(t, Sync(env.syncOpts()))
}

func TestValidateHarnesses(t *testing.T) {
	// Empty harnesses
	err := ValidateHarnesses([]Harness{}, "/some/path")
	require.Error(t, err)
	require.Contains(t, err.Error(), "no harnesses configured")

	// Invalid harness
	err = ValidateHarnesses([]Harness{"invalid"}, "/some/path")
	require.Error(t, err)
	require.Contains(t, err.Error(), "unsupported harness")

	// Valid harnesses
	require.NoError(t, ValidateHarnesses([]Harness{HarnessClaude}, "/some/path"))
	require.NoError(t, ValidateHarnesses([]Harness{HarnessClaude, HarnessOpencode}, "/some/path"))
}
